### DISCUSSION TITLE
How to change order of commands?

### DISCUSSION THREAD

**Discussion by ealap** [topgrade-rs/topgrade#372] [Q&A]
I want to re-order the execution because `cargo` sometimes would take a long time to execute and the custom commands which might require `sudo` only gets executed after the sudo timeout has elapsed.

**Comment by s34m** [comment]
The only way currently possible would be that you manually change the order in the code.

**Reply by agoodshort** [reply]
Looks like there is #376 opened for this.

**Comment by agoodshort** [comment]
For such scenario it feels like `pre_sudo` should run before each commands and not just after the `pre_commands`. By this way your cached credentials should not have the time to expire (the usual 5 minutes by default).
There could be a `sudo -k` at the end to invalidate the cached credentials.

**Reply by SteveLauC** [reply]
> `pre_sudo` should run before `pre_commands`

Good catch!

> There could be a sudo -k at the end to invalidate the cached credentials.

Yeah, this is better!

**Reply by agoodshort** [reply]
Hey @ealap there's a PR with a potential fix on #565 
I didn't have a chance to test it yet, and I'm actually not even facing your issue so I'd first need to reproduce your scenario. Do you want to give it a go and provide feedback?

**Reply by ealap** [reply]
Hi @agoodshort, this is great! I'll try to find time to test it out.

If you want to test it out yourself, you could try install one or more cargo packages with outdated versions, then make sure it is followed by a supported command that needs sudo.

**Reply by SteveLauC** [reply]
Well, I don't let you guys down, but that PR has done ONLY 2 things:

1. move `pre_sudo` before `pre_commands`
2. Clear the cached credential with `sudo -k`

Changing the order of steps is still unimplemented and impossible with the current approach used by Topgrade

**Comment by GideonBear** [comment]
Replaced by #1049

**Comment by GideonBear** [comment]
Ref #1471
